shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Snap.hs

```diff
@@ -44,28 +49,41 @@ data SnapPredicateFailure era -- No predicate failures
 instance NoThunks (SnapPredicateFailure era)
 
 instance (UsesTxOut era, UsesValue era) => STS (SNAP era) where
-  type State (SNAP era) = SnapShots (Crypto era)
-  type Signal (SNAP era) = ()
+  type State (SNAP era) = SnapShots era
+  type Signal (SNAP era) = SlotNo
   type Environment (SNAP era) = LedgerState era
   type BaseM (SNAP era) = ShelleyBase
   type PredicateFailure (SNAP era) = SnapPredicateFailure era
   initialRules = [pure emptySnapShots]
   transitionRules = [snapTransition]
 
 snapTransition ::
-  ( UsesValue era,
-    HasField "address" (Core.TxOut era) (Addr (Crypto era))
-  ) =>
+  (UsesValue era, UsesTxOut era) =>
   TransitionRule (SNAP era)
 snapTransition = do
-  TRC (lstate, s, _) <- judgmentContext
-
-  let LedgerState (UTxOState utxo _ fees _) (DPState dstate pstate) = lstate
-      stake = stakeDistr utxo dstate pstate
+  TRC (lstate, s, slotno) <- judgmentContext
+  let LedgerState (UTxOState (utxo@(UTxO u)) _ fees _) (DPState dstate pstate) = lstate
+  num_blocks_estimate <- liftSTS $ blocksPerEpoch slotno
+  let pulseSize = (div (Map.size u) (fromIntegral num_blocks_estimate)) + 1
+      stake = makePulsingStakeDistr pulseSize utxo dstate pstate
   pure $
     s
       { _pstakeMark = stake,
-        _pstakeSet = _pstakeMark s,
+        _pstakeSet = runToCompletion (_pstakeMark s),
         _pstakeGo = _pstakeSet s,
         _feeSS = fees
       }
+
+-- | Compute an (under) estimate of the number of blocks in the current Epoch, from the current SlotNo.
+--   We want the estimate to be a little too small, as this increases the probability that the
+--   pulsing will be Completed before the end of the Epoch. When we divide this 'too small' number into
+--   the UTxO size, we will get a slighty 'too large' pulseSize, so we should finish early.
+blocksPerEpoch :: SlotNo -> BaseM (SNAP era) Word64
```

Comment:
I would prefix `approx` to this function, to make it clear that it neither predicts nor accurately reports from history the number of blocks in a given epoch.

Is there a reason it takes the slot rather than the epoch (which would seem more usual?)

Reply:
Changing the name is a good idea. The Slotting module needs the SlotNo to compute this. I don't know why.

Reply:
It is now called  approxBlocksPerEpoch, and I have elaborated in the comment how it is used.